pil-analyzer/tests/types.rs

```diff
@@ -561,6 +561,122 @@ fn empty_conditional() {
     type_check(input, &[]);
 }
 
+#[test]
+fn defined_trait() {
+    let input = "
+    trait Add<T> {
+        add: T, T -> T,
+    }
+    impl<T> Add<int> {
+        add: |a, b| a + b,
+    }
+    let r: int = Add::add(3, 4);
+    ";
+    type_check(input, &[("r", "", "int")]);
+}
+
+#[test]
+#[should_panic = "Trait Add not found"]
+fn undefined_trait() {
+    let input = "
+    impl<T> Add<int, int> {
+        add: |a, b| a + b,
+    }
+    ";
+    type_check(input, &[]);
+}
+
+#[test]
+#[should_panic = "Impls for Add: Types (int, fe) and (int, fe) overlap"]
+fn duplicated_trait() {
+    let input = "
+    trait Add<T, Q> {
+        add: T, T -> Q,
+    }
+    impl<T, Q> Add<int, fe> {
+        add: |a, b| a + b,
+    }
+    impl<T, Q> Add<int, fe> {
+        add: |a, b| b + a,
+    }
+    ";
+    type_check(input, &[]);
+}
+
+#[test]
+#[should_panic = "Trait Add has 2 type vars, but implementation has 1"]
+fn impl_with_diff_length() {
+    let input = "
+    trait Add<T, Q> {
+        add: T, T -> Q,
+    }
+    impl<T> Add<int> {
+        add: |a, b| a + b,
+    }
+    ";
+    type_check(input, &[]);
+}
+
+#[test]
+fn impl_combined_test() {
+    let input = "
+    namespace std::convert(4);
+        let fe = || fe();
+    namespace F(4);
+        trait Add<T, Q> {
+            add: T, T -> Q,
+        }
+        impl<T, Q> Add<int, Q> {
+            add: |a, b| std::convert::fe(a + b),
+        }
+        
+        let x: int -> fe = |q| match Add::add(q, 4) {
+            v => v,
+        };
+        let res: fe = x(5);
+    ";
+
+    type_check(input, &[("F.res", "", "fe")]);
+}
+
+#[test]
+fn impl_type_resolution() {
+    let input = "
+    namespace std::convert(4);
+        let fe = || fe();
+    namespace F(4);
+    
+        trait Add<T> {
+            add: T, T -> T,
+        }
+
+        impl<T> Add<fe> {
```

Comment:
Shouldn't this be `impl Add<fe>`? It should be an error if the impl introduces a type variable and does not use it.

Reply:
Added a check about this: https://github.com/powdr-labs/powdr/blob/b175ff3eb91a6ac3f9ad4719e54716a427fa1471/pil-analyzer/src/traits_processor.rs#L66